openssl/src/base64.rs

```diff
@@ -0,0 +1,170 @@
+use std::ptr;
+use std::slice;
+
+use ffi;
+use error::ErrorStack;
+use {cvt_p, cvt};
+
+struct Bio(*mut ffi::BIO);
+
+impl Bio {
+    pub fn as_ptr(&self) -> *mut ffi::BIO {
+        self.0
+    }
+}
+
+impl Drop for Bio {
+    fn drop(&mut self) {
+        unsafe {
+            ffi::BIO_free_all(self.0);
+        }
+    }
+}
+
+/// Encode `data` in base64. If `newlines` is true then the data is
+/// encoded with line breaks.
+pub fn encode(data: &[u8], newlines: bool) -> Result<Vec<u8>, ErrorStack> {
+    if data.is_empty() {
+        return Ok(Vec::new());
+    }
+
+    let bio = unsafe { try!(cvt_p(ffi::BIO_new(ffi::BIO_f_base64()))) };
+
+    let bio = Bio(bio);
+
+    if !newlines {
+        unsafe {
+            ffi::BIO_set_flags(bio.as_ptr(), ffi::BIO_FLAGS_BASE64_NO_NL);
+        }
+    }
+
+    unsafe {
+        // Add some memory to hold the result
+        let mem = try!(cvt_p(ffi::BIO_new(ffi::BIO_s_mem())));
+
+        try!(cvt_p(ffi::BIO_push(bio.as_ptr(), mem)));
+
+        // Write `data` to the BIO
+        try!(cvt(ffi::BIO_write(bio.as_ptr(),
+                                data.as_ptr() as *const _,
+                                data.len() as _)));
+
+        // Flush output
+        try!(cvt(ffi::BIO_flush(bio.as_ptr())));
+
+        // Build a slice from the output
+        let mut ptr = ptr::null_mut();
+        let len = ffi::BIO_get_mem_data(bio.as_ptr(), &mut ptr);
+
+        let s =
+            slice::from_raw_parts(ptr as *const _ as *const _, len as usize);
+
+        Ok(s.to_owned())
+    }
+}
+
+/// Decode `data` from base64. If `nl` is true then the decoder will
+/// accept newlines in the base64 data.
+pub fn decode(data: &[u8], newlines: bool) -> Result<Vec<u8>, ErrorStack> {
+    if data.is_empty() {
+        return Ok(Vec::new());
+    }
+
+    let bio = unsafe { try!(cvt_p(ffi::BIO_new(ffi::BIO_f_base64()))) };
+
+    let bio = Bio(bio);
+
+    unsafe {
+        // Put `data` in a mem_buf
+        let mem_buf = try!(cvt_p(ffi::BIO_new_mem_buf(data.as_ptr() as *const _,
+                                                      data.len() as _)));
+
+        try!(cvt_p(ffi::BIO_push(bio.as_ptr(), mem_buf)));
+
+        if !newlines {
+            ffi::BIO_set_flags(bio.as_ptr(), ffi::BIO_FLAGS_BASE64_NO_NL);
+        }
+    }
+
+    // Allocate output buffer, we know that base64 data will be at
+    // least 4/3 bigger than the raw input
+    let mut data = vec![0; (data.len() * 3 + 3) / 4];
+
+    unsafe {
+        let len = try!(cvt(ffi::BIO_read(bio.as_ptr(),
```

Comment:
The problem is that I don't know how much data I expect precisely, at least not without implementing parts of a base64 decoder just to figure it out. The exact length depends on the number of '=' signs at the end and of course if there are newlines in the input those should be stripped away.

The size of the `vec!` is an upper bound so I know it should all fit in a single read but I'm not sure how to identify short reads.